extensions/positron-python/python_files/positron/positron_ipykernel/data_explorer.py

```diff
@@ -176,12 +193,99 @@ def _get_single_column_schema(self, column_index: int) -> ColumnSchema:
         raise NotImplementedError
 
     def search_schema(self, request: SearchSchemaRequest):
-        return self._search_schema(
-            request.params.search_term,
-            request.params.start_index,
-            request.params.max_results,
+        filters = request.params.filters
+        start_index = request.params.start_index
+        max_results = request.params.max_results
+        if self._search_schema_last_result is not None:
+            last_filters, matches = self._search_schema_last_result
+            if last_filters != filters:
+                matches = self._search_schema_get_matches(filters)
+                self._search_schema_last_result = (filters, matches)
+        else:
+            matches = self._search_schema_get_matches(filters)
+            self._search_schema_last_result = (filters, matches)
+
+        matches_slice = matches[start_index : start_index + max_results]
+        return SearchSchemaResult(
+            matches=TableSchema(columns=matches_slice),
+            total_num_matches=len(matches),
         ).dict()
 
+    def _search_schema_get_matches(self, filters: List[ColumnFilter]) -> List[ColumnSchema]:
+        matchers = self._get_column_filter_functions(filters)
+
+        parent = self
+
+        class SchemaMemo:
+            def __init__(self):
+                self.memo = {}
+
+            def get(self, i):
+                if i in self.memo:
+                    return self.memo[i]
+                else:
+                    schema = parent._get_single_column_schema(i)
+                    self.memo[i] = schema
+                    return schema
+
+        schema_memo = SchemaMemo()
+        matches = []
+        for column_index in range(self.table.shape[1]):
+            # All filters must match
+            no_match = False
+            for matcher in matchers:
+                if not matcher(column_index, schema_memo):
+                    no_match = True
+                    break
+
+            if no_match:
+                continue
+            col_schema = self._get_single_column_schema(column_index)
+            matches.append(col_schema)
```

Comment:
Nit/suggestion - may be personal preference

```suggestion
        matches = [
            self._get_single_column_schema(column_index)
            for column_index in range(self.table.shape[1])
            if all(matcher(column_index, schema_memo) for matcher in matchers)
        ]
```